base/broadcast.jl

```diff
@@ -316,6 +311,20 @@ end
     end
 end
 @inline broadcast_c(f, ::Type{Any}, a...) = f(a...)
+broadcast_c(f, ::Type{Tuple}, As...) =
+    ntuple(k -> f(_tuplebroadcast_getargs(As, k)...), _tuplebroadcast_reslength(As))
+broadcast_c{T}(f, ::Type{Tuple}, ::Type{T}, As...) =
+    ntuple(k -> f(T, _tuplebroadcast_getargs(As, k)...), _tuplebroadcast_reslength(As))
+@inline _tuplebroadcast_getargs(::Tuple{}, k) = ()
+@inline _tuplebroadcast_getargs(As, k) =
+    (_broadcast_getindex(first(As), k), _tuplebroadcast_getargs(tail(As), k)...)
+@noinline _tuplebroadcast_reslength(As) =
```

Comment:
> If inlining the length is the problem, then just defining
> `@noinline _tuplebroadcast_reslength(As) = length(broadcast_indices(As...)[1])`
> instead of these should do (it might be even faster).

Inlining of the length determination code was most but not all of the problem. The length determination code in this pull request is substantially faster than the `length(broadcast_indices(...))` construction. For example,
```julia
using Base.tail
using Base.Broadcast: _broadcast_getindex, broadcast_indices

@inline _tuplebroadcast_getargs(::Tuple{}, k) = ()
@inline _tuplebroadcast_getargs(As, k) =
    (_broadcast_getindex(first(As), k), _tuplebroadcast_getargs(tail(As), k)...)

tuplebc(f, ::Type{Tuple}, As...) =
    ntuple(k -> f(_tuplebroadcast_getargs(As, k)...), _tuplebroadcast_reslength(As))
@noinline _tuplebroadcast_reslength(As) =
    _tuplebroadcast_maxlength(_tuplebroadcast_length(first(As)), tail(As))
@inline _tuplebroadcast_maxlength(l, As) =
    _tuplebroadcast_maxlength(max(l, _tuplebroadcast_length(first(As))), tail(As))
@inline _tuplebroadcast_maxlength(l, ::Tuple{}) = l
@inline _tuplebroadcast_length(t::Tuple) = length(t)
@inline _tuplebroadcast_length(s) = 1

tuplebc2(f, ::Type{Tuple}, As...) =
    ntuple(k -> f(_tuplebroadcast_getargs(As, k)...), _tuplebroadcast_reslength2(As))
@noinline _tuplebroadcast_reslength2(As) = length(broadcast_indices(As...)[1])

tup = (1., 2., 3.)

using BenchmarkTools

@benchmark tuplebc(+, Tuple, $tup, $tup)
@benchmark tuplebc2(+, Tuple, $tup, $tup)
```
yields
```julia
julia> @benchmark tuplebc(+, Tuple, $tup, 1, $tup, 1, $tup, 1, $tup)
BenchmarkTools.Trial:
  memory estimate:  32 bytes
  allocs estimate:  1
  --------------
  minimum time:     36.073 ns (0.00% GC)
  median time:      37.772 ns (0.00% GC)
  mean time:        43.407 ns (7.76% GC)
  maximum time:     3.230 μs (95.74% GC)
  --------------
  samples:          10000
  evals/sample:     993
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark tuplebc2(+, Tuple, $tup, 1, $tup, 1, $tup, 1, $tup)
BenchmarkTools.Trial:
  memory estimate:  32 bytes
  allocs estimate:  1
  --------------
  minimum time:     70.027 ns (0.00% GC)
  median time:      70.797 ns (0.00% GC)
  mean time:        75.237 ns (0.00% GC)
  maximum time:     523.711 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     976
  time tolerance:   5.00%
  memory tolerance: 1.00%
```
i.e. the length determination code in this pull request reduces runtime by roughly a factor of two. Best!